Adds zero quickstart draft (#1326)
pomerium/documentation: content/docs/quickstart.mdx

- Create a [**free Pomerium Zero account**](https://console.pomerium.app/create-account)

## Get Pomerium Zero configuration

In the onboarding console, select the **Docker** tab and copy the Docker Compose configuration.

Paste the configuration in a `docker-compose.yaml` file:

```yaml title="docker-compose.yaml"
services:
  pomerium:
    image: pomerium/pomerium:v0.25.1
    ports:
      - 80:80
      - 443:443
    restart: always
    environment:
    // highlight-start
      POMERIUM_ZERO_TOKEN: <REPLACE_WITH_YOUR_TOKEN>
    // highlight-end
      XDG_CACHE_HOME: /var/cache
      AUTOCERT_DIR: /var/cache/autocert
    volumes:
      - pomerium-cache:/var/cache

volumes:
  pomerium-cache:
```

Replace the value of `POMERIUM_ZERO_TOKEN` with the one generated for you in the onboarding config. The token links this **Pomerium Zero Cluster** with your account.

<details>
  <summary>Clusters Overview</summary>
  <div>Include high-level summary of Clusters concept in PZ. Link to Clusters documentation.</div>
</details>

## Run Docker Compose

```bash
docker compose up -d
```

Pomerium Zero will attempt to detect your **public IP** address to connect you to your cluster.

> But, it won't be able to because you're running Pomerium Zero in an isolated container with its own private network and IP address.

Reviewer: I think right after this we need to showcase what someone is seeing so it's spelled out for them

"You'll see an error like the following (see attached image), and that's completely normal!"

## Apply custom IP address

To connect Pomerium Zero to your cluster, you need to enter a **custom IP address** to override your detected one.

If you're testing with `localhost`, enter `127.0.0.1`, or enter your preferred IP address. You can also enter your preferred listening port.

Select **Apply** to connect to your cluster:

![Connecting to the Pomerium Zero cluster instance in the onboarding screen](./img/quickstart/pz-quickstart-connection-ping.png)

Once you've connected, select **Finish**.

## Connect to upstream services

In the Zero Console, we've provisioned two routes for you with the following names:

- **SSH**
- **Verify**

### Connect to SSH

In a terminal, enter the following command and replace `YOUR_CLUSTER_DOMAIN_NAME` with your own:

```bash

> pomerium-cli tcp ssh.<YOUR_CLUSTER_DOMAIN_NAME>.app:22

Reviewer: Mild nit but it doesn't have the same highlight

</TabItem>
<TabItem value="yum" label="Yum">

```bash
[pomerium-pomerium]
name=pomerium-pomerium
baseurl=https://dl.cloudsmith.io/public/pomerium/pomerium/rpm/el/$releasever/$basearch
repo_gpgcheck=1
enabled=1
gpgkey=https://dl.cloudsmith.io/public/pomerium/pomerium/gpg.6E388440B94E1407.key
gpgcheck=1
sslverify=1
pkg_gpgcheck=1
```

</TabItem>
<TabItem value="deb" label="Deb">

```bash
curl -1sLf 'https://dl.cloudsmith.io/public/pomerium/pomerium/gpg.6E388440B94E1407.key' | apt-key add -
echo "deb https://dl.cloudsmith.io/public/pomerium/pomerium/deb/debian buster main" > /etc/apt/sources.list.d/pomerium-pomerium.list
```

</TabItem>
<TabItem value="binaries" label="Binaries">

```bash
ARCH=[your arch]
OS=[your os]
VERSION=[desired version]
curl -L https://github.com/pomerium/cli/releases/download/${VERSION}/pomerium-cli-${OS}-${ARCH}.tar.gz \
    | tar -z -x
```

</TabItem>
<TabItem value="homebrew" label="Homebrew">

```bash
brew tap pomerium/tap
brew install pomerium-cli
```

</TabItem>
<TabItem value="source" label="Source">

```bash
git clone git@github.com:pomerium/cli.git
cd pomerium
make build
./bin/pomerium-cli --help
```

</TabItem>

</Tabs>

:::

### Connect to Verify

> In the Zero Console, select the **From** route for the Verify app.

Reviewer: ```suggestion
Now, let's test a connection to see if it works. In the Zero Console, select the **From** route for the Verify app.
```

For example:

![Selecting the external URL for the Verify route in the Zero Console](./img/quickstart/pz-quickstart-verify-route.png)

:::note

> Because the Verify service uses a publically accessible URL, Pomerium Zero won't be able to connect to the `jwks.json` endpoint to fetch the signing key used to verify the user's JWT.

Reviewer: ```suggestion
You will see the following error because the Verify service uses a publicly accessible URL. In the current setup, Pomerium Zero won't be able to connect to the `jwks.json` endpoint to fetch the signing key used to verify the user's JWT.
```

![Identity verification failed error when accessing the Verify service](./img/quickstart/pz-identity-verification-failed.png)

:::

## Build your own policy

> In the **Policy** tab, create a new policy that instructs Pomerim to grant access only if the user's email address ends the specified value.

Reviewer: ```suggestion
In the **Policy** tab, create a new policy that instructs Pomerim to grant access only if the user's email address ends with the specified value.
```

![Building a basic policy in Pomerium Zero](./img/quickstart/pz-my-policy.png)

### Attach the policy to a route

Attach your new policy to the Verify route:

![Attach the new policy to the Verify route in the Zero Console](./img/quickstart/pz-attach-policy.png)

Now, if you authenticate using an email address with a domain other than the one specified in the policy, Pomerium Zero will deny the user access.

## Next Steps

Now that you've completed the Pomerium Zero Quickstart, build your first route:

> - **Pomerium Zero Fundamentals: Build Routes**

Reviewer: This doesn't currently link to anything. Are we waiting on the fundamentals to be built too?

Author: Yes, we are.

  ]
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Pomerium Zero Quickstart

The Zero Quickstart shows you how to install and run Pomerium Zero in a Docker container.

## Before you start

- Install [Docker](https://docs.docker.com/install/) and [Docker Compose](https://docs.docker.com/compose/install/)
- Create a (free) [**Pomerium Zero account**](https://console.pomerium.app/create-account)

## Get Pomerium Zero configuration

> In the Zero onboarding screen, select the **Docker** tab and copy the Docker Compose configuration. (If you selected **Finish** before copying the Docker configuration, we've provided a copy below.)

Reviewer 2: It's probably worth mentioning something here about how you get to the onboarding screen, maybe something like "after creating an account, you should see an onboarding page"?

Author: Good point, I'll add that in.

Paste the configuration in a `compose.yaml` file:

```yaml title="compose.yaml" {8,15} showLineNumbers
services:
  pomerium:
    image: pomerium/pomerium:v0.27.0
    ports:
      - 443:443
    restart: always
    environment:
      POMERIUM_ZERO_TOKEN: <CLUSTER_TOKEN>
      XDG_CACHE_HOME: /var/cache
    volumes:
      - pomerium-cache:/var/cache
    networks:
      main:
        aliases:
        - verify.<CLUSTER_STARTER_SUBDOMAIN>.pomerium.app
  verify:
    image: cr.pomerium.com/pomerium/verify:latest
    networks:
      main:
        aliases:
        - verify

networks:
  main: {}

volumes:
  pomerium-cache:
```

Replace the following values:

- **Line 8**: Replace `CLUSTER_TOKEN` with the cluster token generated for you in the onboarding config
- **Line 15**: Replace `CLUSTER_STARTER_SUBDOMAIN` with your generated starter subdomain

:::info Where's my **Starter Domain?**

> You can find your **Starter Domain** at the top of the Zero Console in the navigation bar.

Reviewer 2: I think we don't show the starter domain while on the onboarding page.

Author: We do, since Caleb implemented the change to make the hosted Verify service reachable for deployments running in Docker. Here's the current config in our onboarding flow:
```yaml
services:
  pomerium:
    image: pomerium/pomerium:v0.26.0
    ports:
      - 443:443
    restart: always
    environment:
      POMERIUM_ZERO_TOKEN: AMf-vBy2rzctrTCXBgcAIpUfLKWSme0Bs_HtTcQYb5nbFe8idwaiIt4MjemObphF_L-Pdz_pqCggHHwszIziQU4bC2nGJklhgMTfGGEtpntqfmXRBT6D_m09maHqSJy_75DrLFL2SQOgw3pjuWELm4WhtjD1Rs8aCTl7acjPJe0CKJk5zfSMZcAeFajMhKFpEk-yq_c9G9MN
      XDG_CACHE_HOME: /var/cache
    volumes:
      - pomerium-cache:/var/cache
    networks:
      main:
        aliases:
        - verify.pro-cricket-4498.pomerium.app
  verify:
    image: cr.pomerium.com/pomerium/verify:latest
    networks:
      main:
        aliases:
        - verify

networks:
  main: {}

volumes:
  pomerium-cache:
```

> You can find your **Starter Domain** at the top of the Zero Console in the navigation bar.

Reviewer 2: It might be nice to give some explanation here about what the `verify` app is and why it's included here? (Do we have a docs page about the verify app? Maybe we should add one if not?)

Author: Good point. We do not have anything I can think of that discusses the Verify app. For now, it would be easiest to link to the GH repo and add an explanation about the service and it's intended use.

:::

## Connect to Pomerium Zero

Next, deploy Pomerium Zero with the following command:

```bash
docker compose up -d
```

> This command will deploy a local cluster running Pomerium in a Docker container. If your cluster deploys successfully, the Pomerium Zero cloud will attempt to secure a connection to your cluster.

Reviewer 2: I think we could give some better explanation here. In particular "Pomerium Zero cloud will attempt to secure a connection to your cluster" doesn't make sense to me. Maybe we could say something about how this starts Pomerium in a mode where it connects to the cloud service?

We might also want to hold off on talking about a "cluster" just yet... this might be confusing to some folks if they assume "cluster" implies multiple replicas.

Author: If we want to start being more consistent with terminology, this might be a good time to mention "Zero-managed mode", and that that really means Pomerium connects to the Pomerium Zero cloud?

I see your point with the mention of a cluster. I'm in favor of removing it, we just need to agree on how we want to replace it.

If Pomerium Zero connects successfully, it will notify you in the onboarding screen.

![The successful connection notification in the Pomerium Zero onboarding flow](./img/quickstart/zero-quickstart-connection.png)

Once you've connected, select **Finish**.

:::info

See the [Clusters Concept] page to learn more about clusters and how they work with the Pomerium Zero cloud.

:::

## Connect to upstream services

In the Zero Console, we've provisioned two starter routes, **SSH** and **Verify**, that come with their own DNS records and TLS certificates:

![Viewing the two starter routes](./img/quickstart/pz-quickstart-routes.png)

> ### Connect to SSH

Reviewer 2: I'd recommend starting with the simpler example service first (i.e. swapping this section and the "Connect to Verify" section).

Author: Agree.

In a terminal, enter the following command and replace `CLUSTER_STARTER_DOMAIN` with your own:

```bash
pomerium-cli tcp ssh.<CLUSTER_STARTER_DOMAIN>.app:22
```

A successful request results in a response similar to the one below:

```shell-session
$ pomerium-cli tcp ssh.dominant-cowbird-8176.pomerium.app:22
2024/03/25 16:18:45 listening on 127.0.0.1:52541
```

:::note

If you haven't, install [**Pomerium CLI**](/docs/deploy/clients/pomerium-cli) with any of the following methods:

> <Tabs>

Reviewer 2: I wonder if it makes more sense to link out to a different page for the pomerium-cli installation instructions... it seems a hassle to keep duplicate copies of these instructions in sync.

Author: We can [import Markdown](https://docusaurus.io/docs/markdown-features/react#importing-markdown) in Docusaurus. This would make it so we have one source of truth that we can reuse and maintain.

</TabItem>
<TabItem value="yum" label="Yum">

```bash
[pomerium-pomerium]
name=pomerium-pomerium
baseurl=https://dl.cloudsmith.io/public/pomerium/pomerium/rpm/el/$releasever/$basearch
repo_gpgcheck=1
enabled=1
gpgkey=https://dl.cloudsmith.io/public/pomerium/pomerium/gpg.6E388440B94E1407.key
gpgcheck=1
sslverify=1
pkg_gpgcheck=1
```

</TabItem>
<TabItem value="deb" label="Deb">

```bash
curl -1sLf 'https://dl.cloudsmith.io/public/pomerium/pomerium/gpg.6E388440B94E1407.key' | apt-key add -
echo "deb https://dl.cloudsmith.io/public/pomerium/pomerium/deb/debian buster main" > /etc/apt/sources.list.d/pomerium-pomerium.list
```

</TabItem>
<TabItem value="binaries" label="Binaries">

```bash
ARCH=[your arch]
OS=[your os]
VERSION=[desired version]
curl -L https://github.com/pomerium/cli/releases/download/${VERSION}/pomerium-cli-${OS}-${ARCH}.tar.gz \
    | tar -z -x
```

</TabItem>
<TabItem value="homebrew" label="Homebrew">

```bash
brew tap pomerium/tap
brew install pomerium-cli
```

</TabItem>
<TabItem value="source" label="Source">

```bash
git clone git@github.com:pomerium/cli.git
cd pomerium
make build
./bin/pomerium-cli --help
```

</TabItem>

</Tabs>

:::

### Connect to Verify

> In the Zero Console, select the **From** URL for the **Verify** app.

Reviewer 2: Either above or here, I think this would definitely benefit from some explanation — what is the verify service? what does it demonstrate?

Author: I think it makes most sense to include it at the first mention of the Verify app, so earlier in the page. Will do.

You should land on the Verify page:

![The homepage of the Verify app after the user is redirected](./img/quickstart/pz-identity-verified.png)

> ## Review authorization policy

Reviewer 2: Before talking about policies and certificates, would it make sense to show how to create your own route? I think it might be good to explain how the starter domain is set up so you can add any sub-domain as a route, and the steps involved to set up a new route, i.e.
- pick a sub-domain to use with the From URL
- set up some upstream service to use with the To URL
- attach the demo policy to grant access to the route

Author: Well that was the original intention of the Fundamentals courses. This links to the Routes course first, so they can learn to build a route once they have Zero set up. The Zero Fundamentals course does not re-iterate the Quickstart; it instructs users to go back to the Quickstart and complete it first before continuing with Fundamentals.
